src/machine/machine_rp2040_adc.go

```diff
@@ -20,43 +39,112 @@ func InitADC() {
 	waitForReady()
 }
 
-// Configure configures a ADC pin to be able to be used to read data.
+// Below methods are for standard tinygo, ADC{} pin based access to the ADC peripheral.
+
+// The Configure method sets the ADC pin to analog input mode.
+// Note the ADCConfig parameter is ignored.
+// Does nothing if the GPIO is not an ADC muxed pin.
 func (a ADC) Configure(config ADCConfig) {
+	if c, ok := a.GetADCChannel(); ok {
+		c.Configure(ADCChConfig{})
+	}
+}
+
+// The Get method returns a one-shot ADC sample reading.
+func (a ADC) Get() uint16 {
+	if c, ok := a.GetADCChannel(); ok {
+		return c.GetOnce()
+	}
+	// Not an ADC pin!
+	return 0
+}
+
+// The GetADCChannel method returns the channel associated with the ADC pin.
+// Returns !ok if the pin is not an ADC pin
+func (a ADC) GetADCChannel() (c ADCChannel, ok bool) {
+	ok = true
 	switch a.Pin {
-	case ADC0, ADC1, ADC2, ADC3:
-		a.Pin.Configure(PinConfig{Mode: PinAnalog})
+	case ADC0:
+		c = ADC0_CH
+	case ADC1:
+		c = ADC1_CH
+	case ADC2:
+		c = ADC2_CH
+	case ADC3:
+		c = ADC3_CH
 	default:
-		// invalid ADC
-		return
+		// Invalid ADC pin
+		ok = false
 	}
+	return c, ok
 }
 
-func (a ADC) Get() uint16 {
-	rp.ADC.CS.SetBits(uint32(a.getADCChannel()) << rp.ADC_CS_AINSEL_Pos)
+// Below methods are rp2040 specific, channel based access to the ADC peripheral.
+// The adcChannel methods provide access to the temperature sensor. As well as the pin based channels.
+
+// The Configure method sets the channel's associated pin to analog input mode.
+// Configuring the ADC_TEMP_CH channel will power on the temperature sensor.
+// The powered on temperature sensor increases ADC_AVDD current by approximately 40 μA.
+func (c ADCChannel) Configure(config ADCChConfig) {
+	if p, ok := c.Pin(); ok {
+		p.Configure(PinConfig{Mode: PinAnalog})
+	}
+	if c == ADC_TEMP_CH {
+		// Enable temperature sensor bias source
+		rp.ADC.CS.SetBits(rp.ADC_CS_TS_EN)
+	}
+}
+
+// The GetOnce method returns a one-shot ADC sample reading
+func (c ADCChannel) GetOnce() uint16 {
+	// Make it safe to sample multiple ADC channels in separate go routines.
+	adcLock.Lock()
+	rp.ADC.CS.ReplaceBits(uint32(c), 0b111, rp.ADC_CS_AINSEL_Pos)
 	rp.ADC.CS.SetBits(rp.ADC_CS_START_ONCE)
 
 	waitForReady()
+	adcLock.Unlock()
+
+	// rp2040 is a 12-bit ADC, scale raw reading to 16-bits.
+	return uint16(rp.ADC.RESULT.Get()) << 4
+}
 
-	// rp2040 uses 12-bit sampling, so scale to 16-bit
-	return uint16(rp.ADC.RESULT.Get() << 4)
+// The GetVoltage method does a one-shot sample and returns the reading as a voltage between 0.0 and 3.3V
+func (c ADCChannel) GetVoltage() (volts float32) {
+	return float32(c.GetOnce()>>4) * float32(3.3) / float32(1<<12)
 }
 
+// The GetTemp method does a one-shot sample of the internal temperature sensor.
+// Only works on the ADC_TEMP_CH channel. aka AINSEL=4. Other channels will return 0.0.
+func (c ADCChannel) GetTemp() (celsius float32) {
```

Comment:
It would be good if this function matched the name/signature of other temperature functions such as those in the drivers repo. See https://github.com/tinygo-org/drivers/blob/5741ceb9d179dab8b88e290fda2ad5486345303c/mag3110/mag3110.go#L52

Reply:
Based on the comments. It might be best if I rename ADC_TEMP_CH to TEMP_SENSOR and make it the only public channel. This ensures the pin based channels are only accessible via the standard ADC type methods. And the temperature sensor would be accessed like:
```
ts := machine.TEMP_SENSOR
ts.Configure(machine.ADCConfig{})
milliDegrees int32, err := ts.ReadTemperature()
```
Does that sound alright?
The alternative is to write a separate device driver for the temperature sensor but IMO it's better to just treat it as another ADC channel. The ADC peripheral doesn't distinguish it from the other input channels and it will be simpler to implement the more advanced ADC features if it's not treated as a separate device.

Reply:
That makes sense. No need for separate driver, my point was to have the same interface.